Issue #256 | Try/Catch FlowNFT Hooks

0xdavinchee opened the pull request:
See the issue here: https://github.com/superfluid-finance/superfluid-stories/issues/256

## Summary of Changes
- This adds the try/catch back to the flow nft hooks in the CFA.
- Create a new internal library to encapsulate some of the logic when reverting out of gas in the catch block
- Uncomment foundry coverage sections in ci.canary to prevent breaking changes
- Add test to check if `updateCode` reverts when called by non-host for FlowNFTBase
- Use `getTokenId` instead of repeating token id logic in test code: `helper_Get_NFT_ID` (also for coverage)

0xdavinchee committed:
try/catch as it was

hellwolf commented on packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:
```diff
@@ -470,12 +470,21 @@ contract ConstantFlowAgreementV1 is
+// If the CFA hook actually runs out of gas, not just hitting the safety gas limit, we revert the whole transaction.
+// This solves an issue where the gas estimaton didn't provide enough gas by default for the CFA hook to succeed.
+// See https://medium.com/@wighawag/ethereum-the-concept-of-gas-and-its-dangers-28d0eb809bb2
+            if (gasleft() <= gasLeftBefore / 63) {
```
We should finally create a library function for this, e.g.

```
function _revertWhenOutOfGas() internal {
           if (gasleft() <= gasLeftBefore / 63) {
                revert CFA_HOOK_OUT_OF_GAS();
            }
}
```

0xdavinchee replied:
addressed, added a `SafeGasLibrary` lib, open to any new name suggestions

0xdavinchee committed:
out of gas lib

0xdavinchee committed:
coverage fixes

- uncomment lcov coverage code in ci.canary for nft contracts
- get full coverage to ensure updating code via non-host reverts

0xdavinchee committed:
Merge branch 'dev' into try-catch-cfa-hooks

0xdavinchee committed:
temp test

0xdavinchee committed:
Merge branch 'try-catch-cfa-hooks' of github.com:superfluid-finance/protocol-monorepo into try-catch-cfa-hooks

0xdavinchee committed:
remove temp test code

hellwolf commented on .github/workflows/ci.canary.yml:
```diff
@@ -135,15 +135,15 @@ jobs:
-      #     lcov -e lcov.info -o lcov.info 'packages/ethereum-contracts/contracts/superfluid/*NFT*.sol'
-      #     # merge hardhat and forge coverage files
-      #     lcov -a lcov.info -a packages/ethereum-contracts/coverage/lcov.info -o packages/ethereum-contracts/coverage/lcov.info
+      - name: Install lcov
```
will it work?

0xdavinchee replied:
did another test here: https://github.com/superfluid-finance/protocol-monorepo/actions/runs/4531480256/jobs/7981669257
previously spent a long time trying to get it to work, but just wanted to double check in ci.feature above

0xdavinchee committed:
Merge branch 'dev' into try-catch-cfa-hooks
